chain/network/src/network_protocol/proto_conv.rs

```diff
@@ -0,0 +1,528 @@
+    Handshake, HandshakeFailureReason, PeerMessage, RoutingSyncV2, RoutingTableUpdate,
+};
+use anyhow::{bail, Context};
+/// Contains protobuf <-> network_protocol conversions.
+use borsh::{BorshDeserialize as _, BorshSerialize as _};
+use near_network_primitives::types::{
+    Edge, PartialEdgeInfo, PeerChainInfoV2, PeerInfo, RoutedMessage,
+};
+use near_primitives::block::{Block, BlockHeader, GenesisId};
+use near_primitives::challenge::Challenge;
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::{AnnounceAccount, PeerId};
+use near_primitives::syncing::{EpochSyncFinalizationResponse, EpochSyncResponse};
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::types::EpochId;
+
+impl TryFrom<&proto::CryptoHash> for CryptoHash {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::CryptoHash) -> anyhow::Result<Self> {
+        CryptoHash::try_from(&p.hash[..]).map_err(|err| anyhow::Error::msg(err.to_string()))
+    }
+}
+
+impl TryFrom<&CryptoHash> for proto::CryptoHash {
+    type Error = anyhow::Error;
+    fn try_from(x: &CryptoHash) -> anyhow::Result<Self> {
+        Ok(Self { hash: x.0.into() })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::GenesisId> for GenesisId {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::GenesisId) -> anyhow::Result<Self> {
+        Ok(Self {
+            chain_id: p.chain_id.clone(),
+            hash: (|| CryptoHash::try_from(p.hash.as_ref().context("missing")?))()
+                .context("hash")?,
+        })
+    }
+}
+
+impl TryFrom<&GenesisId> for proto::GenesisId {
+    type Error = anyhow::Error;
+    fn try_from(x: &GenesisId) -> anyhow::Result<Self> {
+        Ok(Self { chain_id: x.chain_id.clone(), hash: Some((&x.hash).try_into().context("hash")?) })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::PeerChainInfo> for PeerChainInfoV2 {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::PeerChainInfo) -> anyhow::Result<Self> {
+        Ok(Self {
+            genesis_id: (|| GenesisId::try_from(p.genesis_id.as_ref().context("missing")?))()
+                .context("genesis_id")?,
+            height: p.height,
+            tracked_shards: p.tracked_shards.clone(),
+            archival: p.archival,
+        })
+    }
+}
+
+impl TryFrom<&PeerChainInfoV2> for proto::PeerChainInfo {
+    type Error = anyhow::Error;
+    fn try_from(x: &PeerChainInfoV2) -> anyhow::Result<Self> {
+        Ok(Self {
+            genesis_id: Some((&x.genesis_id).try_into().context("genesis_id")?),
+            height: x.height,
+            tracked_shards: x.tracked_shards.clone(),
+            archival: x.archival,
+        })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::PublicKey> for PeerId {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::PublicKey) -> anyhow::Result<Self> {
+        Ok(Self::try_from_slice(&p.borsh)?)
```

Comment:
Missing context?

Reply:
This call is so trivial that IMO there is no extra context to add that could be useful (i.e. it would rather obscure the information in the cause chain). The proto message types that I introduced solely for wrapping borsh should rather be treated just like one-field tuple types in rust / newtype in Haskell - extra information to the type system, but not really for the user.